sematic/external_resource.py

```diff
@@ -0,0 +1,229 @@
+# Standard Library
+import logging
+import time
+import uuid
+from dataclasses import dataclass, field
+from enum import Enum, unique
+from typing import final
+
+# Sematic
+from sematic.future_context import AbstractExternalResource
+from sematic.utils.exceptions import IllegalStateTransitionError
+
+logger = logging.getLogger(__name__)
+
+
+@unique
+class ResourceState(Enum):
+    """Represents the state of an external resource.
+
+    Attributes
+    ----------
+    CREATED:
+        The object representing the resource has been instantiated in-memory
+        but no logic to allocate the resource has been executed.
+    ACTIVATING:
+        The resource is being allocated. This may take time. The resource cannot
+        be used yet.
+    ACTIVE:
+        The resource is allocated and ready for use.
+    DEACTIVATING:
+        The resource is being deactivated, and likely is no longer usable.
+    DEACTIVATED:
+        The resource has been deactivated and may not even exist anymore.
+        It is not usable.
+    """
+
+    CREATED = "CREATED"
+    ACTIVATING = "ACTIVATING"
+    ACTIVE = "ACTIVE"
+    DEACTIVATING = "DEACTIVATING"
+    DEACTIVATED = "DEACTIVATED"
+
+    def is_allowed_transition(self, other_state: "ResourceState") -> bool:
```

Comment:
Not for this PR, but we have now 3 state machines (futures/runs, resolutions, resources). We may need some abstraction around it for DRY. But not urgent, not critical, not for this PR.